File: backend/internal/models/nav_history.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
)

// NAVHistory represents the net asset value history for a portfolio
type NAVHistory struct {
	PortfolioID uuid.UUID        `json:"portfolio_id" db:"portfolio_id"`
	Timestamp   time.Time        `json:"timestamp" db:"timestamp"`
	NAV         decimal.Decimal  `json:"nav" db:"nav" validate:"required,gte=0"`
	PnL         decimal.Decimal  `json:"pnl" db:"pnl"`
	Drawdown    *decimal.Decimal `json:"drawdown" db:"drawdown"`
	CreatedAt   time.Time        `json:"created_at" db:"created_at"`
	
	// Related data (not stored in database)
	Portfolio *Portfolio `json:"portfolio,omitempty"`
}

// CreateNAVHistoryRequest represents the request to create a new NAV history entry
type CreateNAVHistoryRequest struct {
	PortfolioID uuid.UUID        `json:"portfolio_id" validate:"required"`
	Timestamp   time.Time        `json:"timestamp" validate:"required"`
	NAV         decimal.Decimal  `json:"nav" validate:"required,gte=0"`
	PnL         decimal.Decimal  `json:"pnl"`
	Drawdown    *decimal.Decimal `json:"drawdown,omitempty"`
}

// NAVHistoryResponse represents the NAV history data returned in API responses
type NAVHistoryResponse struct {
	PortfolioID uuid.UUID        `json:"portfolio_id"`
	Timestamp   time.Time        `json:"timestamp"`
	NAV         decimal.Decimal  `json:"nav"`
	PnL         decimal.Decimal  `json:"pnl"`
	Drawdown    *decimal.Decimal `json:"drawdown"`
	CreatedAt   time.Time        `json:"created_at"`
	Portfolio   *Portfolio       `json:"portfolio,omitempty"`
}

// PerformanceMetrics represents calculated performance metrics for a portfolio
type PerformanceMetrics struct {
	TotalReturn       decimal.Decimal  `json:"total_return"`
	TotalReturnPct    decimal.Decimal  `json:"total_return_pct"`
	AnnualizedReturn  *decimal.Decimal `json:"annualized_return,omitempty"`
	MaxDrawdown       *decimal.Decimal `json:"max_drawdown"`
	CurrentDrawdown   *decimal.Decimal `json:"current_drawdown"`
	VolatilityPct     *decimal.Decimal `json:"volatility_pct,omitempty"`
	SharpeRatio       *decimal.Decimal `json:"sharpe_ratio,omitempty"`
	DaysActive        int              `json:"days_active"`
	HighWaterMark     decimal.Decimal  `json:"high_water_mark"`
}

// ToResponse converts a NAVHistory to NAVHistoryResponse
func (n *NAVHistory) ToResponse() *NAVHistoryResponse {
	return &NAVHistoryResponse{
		PortfolioID: n.PortfolioID,
		Timestamp:   n.Timestamp,
		NAV:         n.NAV,
		PnL:         n.PnL,
		Drawdown:    n.Drawdown,
		CreatedAt:   n.CreatedAt,
		Portfolio:   n.Portfolio,
	}
}

// FromCreateRequest creates a NAVHistory from CreateNAVHistoryRequest
func (n *NAVHistory) FromCreateRequest(req *CreateNAVHistoryRequest) {
	n.PortfolioID = req.PortfolioID
	n.Timestamp = req.Timestamp
	n.NAV = req.NAV
	n.PnL = req.PnL
	n.Drawdown = req.Drawdown
	n.CreatedAt = time.Now()
}

// CalculateDrawdown calculates the drawdown from a high water mark
func (n *NAVHistory) CalculateDrawdown(highWaterMark decimal.Decimal) {
	if highWaterMark.GreaterThan(decimal.Zero) && n.NAV.LessThan(highWaterMark) {
		drawdown := n.NAV.Sub(highWaterMark).Div(highWaterMark).Mul(decimal.NewFromInt(100))
		n.Drawdown = &drawdown
	} else {
		// No drawdown if at or above high water mark
		zero := decimal.Zero
		n.Drawdown = &zero
	}
}

// CalculatePerformanceMetrics calculates performance metrics from NAV history
func CalculatePerformanceMetrics(history []NAVHistory, initialInvestment decimal.Decimal) *PerformanceMetrics {
	if len(history) == 0 {
		return &PerformanceMetrics{}
	}
	
	latest := history[len(history)-1]
	
	// Calculate total return
	totalReturn := latest.NAV.Sub(initialInvestment)
	totalReturnPct := decimal.Zero
	if initialInvestment.GreaterThan(decimal.Zero) {
		totalReturnPct = totalReturn.Div(initialInvestment).Mul(decimal.NewFromInt(100))
	}
	
	// Find high water mark and max drawdown
	highWaterMark := initialInvestment
	var maxDrawdown *decimal.Decimal
	
	for _, entry := range history {
		if entry.NAV.GreaterThan(highWaterMark) {
			highWaterMark = entry.NAV
		}
		
		if entry.Drawdown != nil {
			if maxDrawdown == nil || entry.Drawdown.LessThan(*maxDrawdown) {
				maxDrawdown = entry.Drawdown
			}
		}
	}
	
	// Calculate days active
	daysActive := 0
	if len(history) > 1 {
		firstEntry := history[0]
		lastEntry := history[len(history)-1]
		daysActive = int(lastEntry.Timestamp.Sub(firstEntry.Timestamp).Hours() / 24)
	}
	
	metrics := &PerformanceMetrics{
		TotalReturn:    totalReturn,
		TotalReturnPct: totalReturnPct,
		MaxDrawdown:    maxDrawdown,
		DaysActive:     daysActive,
		HighWaterMark:  highWaterMark,
	}
	
	// Calculate current drawdown
	if highWaterMark.GreaterThan(decimal.Zero) && latest.NAV.LessThan(highWaterMark) {
		currentDrawdown := latest.NAV.Sub(highWaterMark).Div(highWaterMark).Mul(decimal.NewFromInt(100))
		metrics.CurrentDrawdown = &currentDrawdown
	}
	
	// Calculate annualized return if we have enough data
	if daysActive > 0 {
		years := decimal.NewFromFloat(float64(daysActive) / 365.25)
		if years.GreaterThan(decimal.Zero) && initialInvestment.GreaterThan(decimal.Zero) {
			// Annualized return = (Final Value / Initial Value)^(1/years) - 1
			finalRatio := latest.NAV.Div(initialInvestment)
			if finalRatio.GreaterThan(decimal.Zero) {
				// Simplified calculation - for more accurate results, would need proper power function
				annualizedReturn := finalRatio.Sub(decimal.NewFromInt(1)).Div(years).Mul(decimal.NewFromInt(100))
				metrics.AnnualizedReturn = &annualizedReturn
			}
		}
	}
	
	return metrics
}

File: frontend/src/types/nav-history.ts
import { z } from 'zod';
import type { Portfolio } from './portfolio';

// NAV History interfaces
export interface NAVHistory {
  portfolio_id: string;
  timestamp: string;
  nav: number;
  pnl: number;
  drawdown?: number;
  created_at: string;
  portfolio?: Portfolio;
}

export interface CreateNAVHistoryRequest {
  portfolio_id: string;
  timestamp: string;
  nav: number;
  pnl: number;
  drawdown?: number;
}

export interface NAVHistoryResponse {
  portfolio_id: string;
  timestamp: string;
  nav: number;
  pnl: number;
  drawdown?: number;
  created_at: string;
  portfolio?: Portfolio;
}

// Performance metrics interface
export interface PerformanceMetrics {
  total_return: number;
  total_return_pct: number;
  annualized_return?: number;
  max_drawdown?: number;
  current_drawdown?: number;
  volatility_pct?: number;
  sharpe_ratio?: number;
  days_active: number;
  high_water_mark: number;
}

// Chart data interfaces for visualization
export interface NAVChartData {
  timestamp: string;
  nav: number;
  pnl: number;
  drawdown?: number;
}

export interface PerformanceChartData {
  date: string;
  nav: number;
  benchmark?: number;
  drawdown: number;
}

// Zod validation schemas
export const createNAVHistorySchema = z.object({
  portfolio_id: z.string().uuid('Portfolio ID must be a valid UUID'),
  timestamp: z.string().datetime('Timestamp must be a valid ISO datetime string'),
  nav: z.number().min(0, 'NAV must be non-negative'),
  pnl: z.number(),
  drawdown: z.number().optional(),
});

// Helper functions for performance calculations
export const calculatePerformanceMetrics = (
  navHistory: NAVHistory[],
  initialInvestment: number
): PerformanceMetrics => {
  if (navHistory.length === 0) {
    return {
      total_return: 0,
      total_return_pct: 0,
      days_active: 0,
      high_water_mark: initialInvestment,
    };
  }

  const latest = navHistory[navHistory.length - 1];
  
  // Calculate total return
  const totalReturn = latest.nav - initialInvestment;
  const totalReturnPct = initialInvestment > 0 ? (totalReturn / initialInvestment) * 100 : 0;
  
  // Find high water mark and max drawdown
  let highWaterMark = initialInvestment;
  let maxDrawdown: number | undefined;
  
  for (const entry of navHistory) {
    if (entry.nav > highWaterMark) {
      highWaterMark = entry.nav;
    }
    
    if (entry.drawdown !== undefined) {
      if (maxDrawdown === undefined || entry.drawdown < maxDrawdown) {
        maxDrawdown = entry.drawdown;
      }
    }
  }
  
  // Calculate days active
  let daysActive = 0;
  if (navHistory.length > 1) {
    const firstEntry = new Date(navHistory[0].timestamp);
    const lastEntry = new Date(navHistory[navHistory.length - 1].timestamp);
    daysActive = Math.floor((lastEntry.getTime() - firstEntry.getTime()) / (1000 * 60 * 60 * 24));
  }
  
  // Calculate current drawdown
  let currentDrawdown: number | undefined;
  if (highWaterMark > 0 && latest.nav < highWaterMark) {
    currentDrawdown = ((latest.nav - highWaterMark) / highWaterMark) * 100;
  }
  
  // Calculate annualized return if we have enough data
  let annualizedReturn: number | undefined;
  if (daysActive > 0) {
    const years = daysActive / 365.25;
    if (years > 0 && initialInvestment > 0) {
      // Simplified annualized return calculation
      const finalRatio = latest.nav / initialInvestment;
      if (finalRatio > 0) {
        annualizedReturn = ((finalRatio - 1) / years) * 100;
      }
    }
  }
  
  return {
    total_return: totalReturn,
    total_return_pct: totalReturnPct,
    annualized_return: annualizedReturn,
    max_drawdown: maxDrawdown,
    current_drawdown: currentDrawdown,
    days_active: daysActive,
    high_water_mark: highWaterMark,
  };
};

// Helper function to format NAV history for charts
export const formatNAVHistoryForChart = (navHistory: NAVHistory[]): NAVChartData[] => {
  return navHistory.map(entry => ({
    timestamp: entry.timestamp,
    nav: entry.nav,
    pnl: entry.pnl,
    drawdown: entry.drawdown,
  }));
};

// Type inference from schemas
export type CreateNAVHistoryFormData = z.infer<typeof createNAVHistorySchema>;